File: modules/ui/src/main/java/eu/europa/ec/leos/web/ui/component/RangeSliderComponent.java
/**
 * Copyright 2016 European Commission
 *
 * Licensed under the EUPL, Version 1.1 or – as soon they will be approved by the European Commission - subsequent versions of the EUPL (the "Licence");
 * You may not use this work except in compliance with the Licence.
 * You may obtain a copy of the Licence at:
 *
 *     https://joinup.ec.europa.eu/software/page/eupl
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the Licence is distributed on an "AS IS" basis,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the Licence for the specific language governing permissions and limitations under the Licence.
 */
package eu.europa.ec.leos.web.ui.component;

import com.vaadin.annotations.JavaScript;
import com.vaadin.annotations.StyleSheet;
import com.vaadin.ui.AbstractJavaScriptComponent;
import com.vaadin.ui.Component;
import com.vaadin.ui.JavaScriptFunction;
import com.vaadin.util.ReflectTools;
import elemental.json.JsonArray;
import eu.europa.ec.leos.web.support.LeosCacheToken;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.Serializable;
import java.lang.reflect.Method;
import java.util.ArrayList;
import java.util.List;
import java.util.Set;

@StyleSheet({"vaadin://../lib/noUiSlider_8.3.0/css/nouislider.css" + LeosCacheToken.TOKEN,
             "vaadin://../lib/noUiSlider_8.3.0/css/nouislider.pips.css" + LeosCacheToken.TOKEN,
             "vaadin://../lib/noUiSlider_8.3.0/css/nouislider.tooltips.css" + LeosCacheToken.TOKEN,
             "vaadin://../assets/css/leos-nouislider.css" + LeosCacheToken.TOKEN})
@JavaScript({"vaadin://../js/web/rangeSliderConnector.js"+ LeosCacheToken.TOKEN})
public class RangeSliderComponent extends AbstractJavaScriptComponent {

    private static final Logger LOG = LoggerFactory.getLogger(RangeSliderComponent.class);

    public RangeSliderComponent() {
        setImmediate(true);
        registerSliderChangeCallback();
    }

    public void setSteps(Set<String> values) {
        getState().stepValues = values;
    }

    public Set<String> getSteps() {
        // state read operation must not trigger a state change event
        return getState(false).stepValues;
    }

    public void setHandles(String[] values) {
        getState().handleValues = values;
    }

    public String[] getHandles() {
        // state read operation must not trigger a state change event
        return getState(false).handleValues;
    }

    @Override
    protected RangeSliderState getState() {
        return (RangeSliderState) super.getState();
    }

    @Override
    protected RangeSliderState getState(boolean markAsDirty) {
        return (RangeSliderState) super.getState(markAsDirty);
    }

    protected void registerSliderChangeCallback() {
        // handle slider changes from client-side
        addFunction("onSliderChange", new JavaScriptFunction() {
            @Override
            public void call(JsonArray args) {
                LOG.debug("Slider change from client-side... {}", args.toJson());
                // FIXME callback comes from client side, but how should state be updated on client side???
                getState(false).handleValues = convertArray(args.getArray(0));
                // FIXME getState() or markAsDirty() should probably be used here,
                // FIXME but will certainly cause change events that will lead to an infinite loop!!!
                // FIXME the SERVER side should probably check if values actually changed to avoid infinite loops
                // FIXME the CLIENT side should probably check if values actually changed to avoid infinite loops
                fireHandlesEvent();
            }
        });
    }

    public void addHandlesListener(HandlesListener listener) {
        addListener(HandlesEvent.class, listener, HandlesListener.VALUES_CHANGED_METHOD);
    }

    public interface HandlesListener extends Serializable {
        Method VALUES_CHANGED_METHOD =
                ReflectTools.findMethod(HandlesListener.class, "valuesChanged", HandlesEvent.class);
        void valuesChanged(HandlesEvent event);
    }

    public static class HandlesEvent extends Component.Event {

        public HandlesEvent(Component source) {
            super(source);
        }

        @Override
        public RangeSliderComponent getComponent() {
            return (RangeSliderComponent) super.getComponent();
        }
    }

    protected void fireHandlesEvent() {
        if (isEnabled() && !isReadOnly()) {
            fireEvent(new HandlesEvent(this));
        }
    }

    protected String[] convertArray(JsonArray input) {
        // FIXME replace with better implementation, maybe using GSON?
        List<String> output = new ArrayList<>();
        for(int i = 0; i < input.length(); i++){
            output.add(input.getString(i));
        }
        return output.toArray(new String[0]);
    }
}

File: modules/js-web/src/main/js/web/rangeSliderComponent.js
/*
 * Copyright 2016 European Commission
 *
 * Licensed under the EUPL, Version 1.1 or – as soon they will be approved by the European Commission - subsequent versions of the EUPL (the "Licence");
 * You may not use this work except in compliance with the Licence.
 * You may obtain a copy of the Licence at:
 *
 *     https://joinup.ec.europa.eu/software/page/eupl
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the Licence is distributed on an "AS IS" basis,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the Licence for the specific language governing permissions and limitations under the Licence.
 */
; // jshint ignore:line
define(function rangeSliderComponentModule(require) {
    "use strict";

    // load module dependencies
    var log = require("logger");
    var noUiSlider = require("nouislider");
    var wNumb = require("wnumb");

    // handle slider initialization
    function _initSlider(connector) {
        var element = connector.getElement();
        var state = connector.getState();
        var steps = state.stepValues;
        var handles = state.handleValues;

        log.debug("Creating slider...", steps, handles);
        var slider = noUiSlider.create(element, {
            orientation: "horizontal",
            connect: true,
            range: _stepRange(steps),
            step: 1,
            start: handles,
            format: wNumb({
                decimals: 0,
                edit: _stepEdit,
                undo: _stepUndo
            }),
            pips: {
                mode: "steps",
                density: 3,
                format: wNumb({
                    edit: _stepEdit
                })
            }
        });

        log.debug("Registering connector unregistration listener...");
        connector.onUnregister = _connectorUnregistrationListener;

        log.debug("Registering connector state change listener...");
        connector.onStateChange = _connectorStateChangeListener;

        log.debug("Registering slider change listener...");
        slider.on("change", function onChange(handleValues) {
            _sliderChangeListener(handleValues, connector);
        });

        function _stepRange(steps) {
            // step index range (0-based)
            return {
                "min": (steps ? 0 : -1),
                "max": (steps ? (steps.length -1) : -1)
            };
        }

        function _stepUndo(value) {
            // pre-decode step value (string -> string)
            // convert step value to step index
            return steps.indexOf(value).toString();
        }

        function _stepEdit(formattedValue, originalValue) {
            // post-encode step value ([string, number] -> string)
            // convert step index to step value
            return steps[originalValue];
        }
    }

    // handle connector unregistration on client-side
    function _connectorUnregistrationListener() {
        var connector = this;
        var element = connector.getElement();
        log.debug("Connector unregistered... destroying slider!");
        element.noUiSlider.destroy();
    }

    // handle connector state change from server-side
    function _connectorStateChangeListener() {
        var connector = this;
        var state = connector.getState();
        var handleValues = state.handleValues;
        var element = connector.getElement();
        log.debug("Connector state changed...", handleValues);
        // FIXME check if values actually changed to avoid infinite loops???
        element.noUiSlider.set(handleValues);
    }

    // handle slider change from client-side
    function _sliderChangeListener(handleValues, connector) {
        log.debug("Slider changed...", handleValues);
        connector.onSliderChange(handleValues);
        // FIXME update shared state to keep synchronization with server-side???
    }

    log.debug("Range Slider Component module is ready!");

    return {
        init: _initSlider
    };
});